### ISSUE TITLE
Model Uncertainty

### ISSUE THREAD

**Issue by mflova** [mattiasegu/uncertainty_estimation_deep_learning#3]
First of all, congratulations for the publication, really interesting one! What I am posting is not an issue of the code, but 2 technical doubts about "Model Uncertainty" that may be easy to answer for you.

I have trained the ResNet18 with dropout (basically a one-output regression modified). Later, when I launched the eval.py script (--p 0.02 --load_model_name resnet18 --test_model_name resnet18_dropout_adf), as expected, I saw how the p value clearly affects the output variance of the network. However, it seems hard to tune this value in order to obtain accurate variances coming from the model uncertainty. 
**Do you think it is a good idea to choose it by minimizing the NLL parameter?**
When it comes about the formula used to calculate NLL in "End-to-End Steering Angle Prediction" from the paper, the NLL is calculated as follows:
`1/2*log(sigma) + 1/(2*sigma)*(ygt-ypred)^2`
Second term is always positive, while the first one can be both positive or negative depending on the value of sigma. Because of this, the result sometimes is positive and sometimes negative, what results in what I think it is a wrong output. **Am I correctly applying this formula?**

Thank you in advance and sorry for the inconveniences.

**Comment by versatran01** [comment]
NLL can be negative, take the exponential will give you > 0 likelihood.

**Comment by mattiasegu** [comment]
Hi @mflova , I am happy that you appreciate our work! As @versatran01 mentioned, negative log-likelihood (NLL) can be negative. If you are looking for something that is strictly non-negative, that is the likehood. However, _maximizing_ the likelihood is equivalent to _minimizing_ the negative likelihood. Moreover, for the sake of optimization, likelihood and log-likelihood are equivalent, with the latter being the default choice because it deals better with numerical optimization issues.

About the equivalence of likelihood and log-likelihood for optimization, see [THIS](https://towardsdatascience.com/probability-concepts-explained-maximum-likelihood-estimation-c7b4342fdbb1#:~:text=The%20log%20likelihood&text=This%20is%20important%20because%20it,instead%20of%20the%20original%20likelihood.) Medium blog post, section 'The log likelihood', where it states that `the maximum value of the log of the probability occurs at the same point as the original probability function`.

For this reason, maximizing the likelihood is in theory equivalent to minimizing the negative log-likelihood. I hope this address your question :)

**Comment by mflova** [comment]
Perfect, thank you very much :)
One last question:
In order to minimise the NLL in eval.py (and thus choosing optimal p), I have trained my ResNet (dropout ON) with really small resulting error. However, with dropout enabled (MC sampling) the outputs of my network oscillates so much that the predictions seems to be almost random, even with really smaller p values in eval.py compared to the ones from training. When training, I have used a dropout value of 0.1 (only in covolutional layers, as default by the ResNet model in your repository. I think this is not a problem since the resnet_dropout_adf model also has same dropout, so the high variance in outputs shouldn't be coming from here).
Have you faced this issue? Would you know which training setup may help on these situations (dropout on fully connected layers, more dropout when training, remove batch normalization..)? I don't know which dropout (layers or % approx) you configured on your experiments in order to make it.

**Comment by mattiasegu** [comment]
Hi @mflova ,

I never faced this issue. When training the model with dropout layers, dropout was placed only on conv layers and with rate 0.2. In that case, the evaluation was conducted with the same drop rate and results were satisfying. Given your description, I think that it is possible that the model was not loaded correctly in eval.py. Did you set the correct flags for the model to load? Are you running the code on GPU or CPU? If the answer is CPU, the recent pull request #4 fixed a bug with model loading in eval.py.

Please let me know if this answers your questions.

**Comment by mflova** [comment]
> Hi @mflova ,
> 
> I never faced this issue. When training the model with dropout layers, dropout was placed only on conv layers and with rate 0.2. In that case, the evaluation was conducted with the same drop rate and results were satisfying. Given your description, I think that it is possible that the model was not loaded correctly in eval.py. Did you set the correct flags for the model to load? Are you running the code on GPU or CPU? If the answer is CPU, the recent pull request #4 fixed a bug with model loading in eval.py.
> 
> Please let me know if this answers your questions.

Problem solved.
As I observed this high variance even with P_training >> P_eval I decided to add more complexity of the model by moving to Resnet18, 34 or even 50, without obtaining results. I did this in order to add more complexity and thus increaseing the redundancy of the model, decreasing the variance of the model uncertainty. However, it seems the problem was in the fully connected layer. Although the dropout was only applied to conv layers, these ones play a very important role when performing the MC sampling. The problem was solved by adding much much more hidden neurons or conexions in this part.
Thank you very much for your early replies, I close this "issue"!

PS: I think I spotted a small mistake in eval.py. At "compute_log_likelihood" the third argument should be torch.sqrt(outputs_variance) instead of outputs_variance.
